### PR TITLE
Adding serde serialization

### LINKED ISSUE #94: Refactor serialization
Refactor the serialization and deserialization of packets. In an ideal world, we should use `serde` to handle all types of serial/deserialization operations if possible.

### PR DESCRIPTION
**Pull request by ryan-summers**
This PR fixes #94 by refactoring serialization to utilize serde.

* A new `src/packets.rs` file has been created to provide rust representations of all MQTT control packets.
* A new `src/ser/serializer.rs` `serde::Serializer` implementation has been added
* A new `src/types.rs` file has been added for MQTT-defined wrapper types

### CODE REVIEW

File: src/ser/serializer.rs [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+const MAX_FIXED_HEADER_SIZE: usize = 5;
+
+/// Errors that result from the serialization process
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Error {
+    /// The provided memory buffer did not have enough space to serialize into.
+    InsufficientMemory,
+
+    /// The requested feature will not be implemented.
+    WontImplement,
```
Comment (jordens) [inline src/ser/serializer.rs:46]: I suspect in general this variant isn't needed and code should `unimplemented!()`
Same for `de`.

File: src/ser/serializer.rs [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    pub fn finalize(self, typ: MessageType, flags: u8) -> Result<&'a [u8], Error> {
+        let len = self.index - MAX_FIXED_HEADER_SIZE;
+
+        let mut buffer = VarintBuffer::new();
+        buffer
+            .write_u32_varint(len as u32)
+            .map_err(|_| Error::InsufficientMemory)?;
+
+        // Write the remaining packet length.
+        self.buf[MAX_FIXED_HEADER_SIZE - buffer.data.len()..][..buffer.data.len()]
```
Comment (jordens) [inline src/ser/serializer.rs:116]:
```suggestion
        self.buf[MAX_FIXED_HEADER_SIZE - buffer.data.len()..MAX_FIXED_HEADER_SIZE]
```

File: src/ser/serializer.rs [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            .copy_from_slice(&buffer.data);
+
+        // Write the header
+        let header: u8 = *0u8.set_bits(4..8, typ as u8).set_bits(0..4, flags);
+        self.buf[MAX_FIXED_HEADER_SIZE - buffer.data.len() - 1] = header;
+
+        Ok(&self.buf[..self.index][MAX_FIXED_HEADER_SIZE - buffer.data.len() - 1..])
```
Comment (jordens) [inline src/ser/serializer.rs:123]:
```suggestion
        Ok(&self.buf[MAX_FIXED_HEADER_SIZE - buffer.data.len() - 1..self.index])
```

File: src/mqtt_client.rs [hunk 4]
```diff
@@ -361,18 +360,22 @@ impl<
...
+            packet_id: Some(id),
+            payload: data,
+            retain,
+            qos,
+            dup: false,
+        };
 
+        crate::info!("Sending: {:?}", publish);
+        let mut buffer: [u8; MSG_SIZE] = [0; MSG_SIZE];
+        let packet = crate::ser::control_packet::to_buffer(&mut buffer, publish)?;
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client.rs:378]: We need `packet` later to pass it to the session state for recording (and possible republication) in this case

File: src/mqtt_client.rs [hunk 5]
```diff
@@ -516,9 +505,15 @@ impl<
...
+                    packet_id: rec.packet_id,
+                    code: 0,
+                    properties: Properties(&[]),
+                };
+
+                crate::info!("Sending: {:?}", pubrel);
                 let mut buffer: [u8; MSG_SIZE] = [0; MSG_SIZE];
-                let packet = serialize::pubrel_packet(&mut buffer, rec.packet_id, 0, &[])?;
-                info!("Sending PubRel({})", rec.packet_id);
+                let packet = crate::ser::control_packet::to_buffer(&mut buffer, pubrel)?;
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client.rs:516]: Same as previous - we need to record the serialized pubrel packet for future retransmission.

**Review by jordens — APPROVED** [review]

**Pull request merged**